Stabilize, optimize, and increase robustness of QuickSort

LilithHafner opened the pull request:
"Quicksort is unstable and attempts to stabilize it hurt performance" (citation needed).

I believe I have simultaneously stabilized and sped up quicksort, which should allow us to simplify sorting policy.

Instability comes from the partition algorithm. We currently use something like this
```julia
while i < j
    while(v[i] < pivot) i += 1 end
    while(v[j] > pivot) j -= 1 end
    v[i], v[j] = v[j], v[i]
end
```
Which is an efficient in place unstable partition algorithm.

If we allow ourselves scratch space to conduct the partition, we can use something like this instead
```julia
for x in v
    if x < pivot
        t[i] = x
        i += 1
    else
        t[j] = x
        j -= 1
    end
end
```
This new approach is still unstable, but it is predictably unstable. Everything before the pivot is stable, and everything after it is reversed. After several nested partitions, we still only need to reverse half the elements one time to reintroduce stability.

[Benchmarks](https://gist.github.com/LilithHafner/56fa5b3e4ed984bcd5af086adbcb98b8) reveal that this approach is typically faster than the original, possibly because the order of comparisons and computations is known at the beginning of the partition rather than depending on the branches during the partition. In effect, the branches in this new algorithm have less impact on future code execution. I have not profiled this implementation at the CPU level and don't know for sure why it is a speedup.

<img width="698" alt="Benchmark Results" src="https://user-images.githubusercontent.com/60898866/167261392-588ed3d7-c891-44c8-ac14-acda77d8c3fd.png">
Caveat 1: I benchmarked on a somewhat inconsistently noisy machine.

Caveat 2: This is sometimes a slowdown. See graphs.

Future work indicated by this PR includes
- Move MergeSort to SortingAlgorithms.jl
- Now that every sorting algorithm in Base's sorting policy except InsertionSort uses scratch space, it becomes even more valuable to effectively pass scratch-space vectors around. For example, in this PR, fpsort! results in redundant scratch space allocations.
- We partition in quicksort and fpsort!. Generalizing partition! and factoring it out should result in substantial readability improvements in fpsort! which, in turn, should facilitate extending support optimized sorting for unions with Missing to all vectors. We could also export `partition!`.

Closes #11429
Closes #42713
Closes #32675

Commit:
Stabilize and optimize QuickSort

petvana commented:
It seems like solving #11429 as a side-effect, but at the cost of losing in-place property.

petvana commented on base/sort.jl:
```diff
@@ -529,75 +530,107 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
-            # stack space in the worst case (rather than O(n))
-            lo < (j-1) && sort!(v, lo, j-1, a, o)
+# Needed for bootstrapping because `view` is not available.
+function reverse_view!(v, lo, hi)
```
Isn't this the same as `reverse!` from `Base`?

LilithHafner replied:
Yes, it is, but we can't use `view` in sort.jl because sorting comes before views in bootstrapping. I've added an additional comment to clarify, and tests to verify that it is true.

petvana replied:
Thanks for the comment, but I don't see the difference between `reverse!(view(v, lo:hi))` and `reverse!(v, lo, hi)`, except for the unused return value. I may miss smth.

LilithHafner replied:
Oh, good catch! I didn't know about the `reverse!(v, start, stop)` function, and thought you were referring to `reverse(view(v, lo, hi))`. Thanks for the link. I've removed the redundancy.

petvana commented on base/sort.jl:
```diff
@@ -529,75 +530,107 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
+    v
+end
+
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort, o::Ordering, t::AbstractVector=similar(v), swap=false, rev=false)
```
Would it make sense to change `rev` to `..., rev::Val(S)=Val(false)) where S` to compile two versions and remove some branching from `partition!`?

LilithHafner replied:
One obstacle to that is that the [`rev = !rev`](https://github.com/LilithHafner/julia/blob/afbed36bfe68d8b66bb977b71fbce19b50b93e22/base/sort.jl#L624) lines would introduce type instability. Nevertheless, I tested making this change and commenting out those (stability tests failed) but I did not measure a performance increase. I imagine the compiler is already doing this or something like it for us. When I added branching based on `rev` into the inner loop of `partition!` I didn't notice any regression, so I doubt the compiler is actually putting a branch there.

petvana replied:
`rev = !rev` can be solved a local variable, like `rev2 = rev == Val(true)` (known at compilation time), but is not necessary if you don't see any regression.

Commit:
Remove redundant reverse_view! thanks @petvana

Commit:
test invalid lt to close #11429

LilithHafner commented:
> It seems like solving #11429 as a side-effect

Good observation! I've added tests to confirm, and indeed it does.

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

petvana commented on base/sort.jl:
```diff
@@ -529,75 +530,99 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
-        if lt(o, v[lo], v[mi])
-            v[mi], v[lo] = v[lo], v[mi]
-        end
+# TODO replace with rand(lo:hi) if Random is ever a part of Base
```
What about hashing, instead of non-deterministic random generator?
``` julia
i1 = mod(hash(lo), hi-lo+1) + lo
i2 = mod(hash(hi), hi-lo+1) + lo
i3 = mod(hash(hi-lo), hi-lo+1) + lo
```

LilithHafner replied:
Good idea!

Benchmarking with
```julia
@btime sort!(x, alg=QuickSort) setup=(x=rand(1:100000, 2000)) evals=1 samples=50000;
```
I get
`42.216 μs (1.00)` for median of `lo`, `mi`, `hi`
`45.960 μs (1.08)` for `rand`
`48.168 μs (1.14)` for median of 3 hashes
`44.711 μs (1.06)` for median of `lo`, `hi`, and `mod(hash(hi), hi-lo-1) + lo + 1`.
But these differences are pretty close to my limit for benchmark accuracy, so I don't know if they are accurate.

I'm still looking into worst-case behavior now.

petvana replied:
Worst cases can be generated quite easily, as in [this example](https://gist.github.com/petvana/8008a1ed4a2d5513f4bb2e398978b695). The hashing cannot eliminate them; however, it seems less probable that any algorithm would produce them accidentally. What I see more interesting is whether such pathological cases can be detected by the number of `while` iterations. It would enable us to call some fallback algorithm, like merge sort, or find more accurate median.

LilithHafner replied:
In this PR, it is even easier to find a quadratic case: `vcat(1:10000, 10000:-1:1)` and `vcat(10000:-1:1, 1:10000)` are both quadratic there but efficient on master because the process of selecting a pivot before the partition and then swapping the pivot in after disrupt the order. That swapping around is unstable. To keep stability and not decrease robustness we must either randomize or hash the majority of the indices used to select pivots (or add worst-case detection and a fallback). Viable options include: 
.1) insert 
```julia
lo = mod(hash(lo), mi-lo) + lo
hi = mod(hash(hi), hi-mi) + mi + 1
```
at the top of the `select_pivot` function to hash 2/3 of the indices.
2) hash all three indices independently as you do above
3) replace the whole select pivot function with `mod(hash(lo), hi-lo+1) + lo` to hash 1/1 of the indices
4) detect worst-case and dispatch to a more robust algorithm

I benchmarked this test function on the current version and options 1-3
```
function test()
    @btime sort!(x, alg=QuickSort) setup=(x=rand(1:100000, 2000)) evals=1 samples=50000;
    @btime sort!(x, alg=QuickSort) setup=(x=sort!(rand(1:100000, 2000))) evals=1 samples=50000;
    @btime sort!(x, alg=QuickSort) setup=(x=vcat(1:10000,10000:-1:1)) evals=1 samples=3000;
    @btime sort!(x, alg=QuickSort) setup=(x=vcat(10000:-1:1, 1:10000)) evals=1 samples=3000;
    nothing
end
```
With these results:
```
As is (1a30832b604f7f153ba3965f799a6f2924e98be0)
  42.039 μs (1 allocation: 15.75 KiB)
  19.532 μs (1 allocation: 15.75 KiB)
  100.561 ms (2 allocations: 156.30 KiB)
  97.124 ms (2 allocations: 156.30 KiB)
1)
  46.403 μs (1 allocation: 15.75 KiB)
  23.195 μs (1 allocation: 15.75 KiB)
  501.198 μs (2 allocations: 156.30 KiB)
  505.714 μs (2 allocations: 156.30 KiB)
2)
  47.160 μs (1 allocation: 15.75 KiB)
  26.952 μs (1 allocation: 15.75 KiB)
  469.629 μs (2 allocations: 156.30 KiB)
  455.273 μs (2 allocations: 156.30 KiB)
3)
  47.103 μs (1 allocation: 15.75 KiB)
  28.547 μs (1 allocation: 15.75 KiB)
  491.788 μs (2 allocations: 156.30 KiB)
  491.719 μs (2 allocations: 156.30 KiB)
```
I favor option 3 because it is the simplest and I don't value the 22% speedup in the presorted case particularly highly because there is already a presorted check in the pipeline.

There are still quadratic worst cases with a hash-chosen pivot, but I think they are similarly unlikely to emerge than the quadratic cases with master's version. It is certainly possible to detect and eliminate quadratic runtime. However, I'd like to call that out of scope for this PR so long as there isn't a regression.

oscardssmith replied:
I think that using a random pivot is strongly worth considering since it prevents the possibility of denial of service attacks.

LilithHafner replied:
Agreed. My preference right now is to move all the sorting algorithms in Sort.jl except for InsertionSort to a stdlib so that they can depend on all of Base and on other stdlibs. Then we would replace `mod(hash(lo), hi-lo+1) + lo` with `rand(lo:hi)`. I think there are other cases where fast sorting algorithms benefit from having more dependencies, and I doubt that we will see much if any regression in build times with only insertion sort.

LilithHafner replied:
Right now we can't use `rand` in Sort.jl because `rand` is defined in a stdlib.

LilithHafner replied:
`rand` is also not cryptographically secure, so to get full protection against a strong adversary we would still need quadratic case detection.

LilithHafner replied:
Update: pivot selection is randomized!

petvana replied:
It may sound too conservative, but wouldn't it be better to detect the quadratic cases and thus abandon randomization (given the (un)expected side effects)?

LilithHafner replied:
Perhaps. Randomization proved harder to implement than I anticipated. However now that it's done, looking back over the total diff, it's pretty small. To implement it takes 2 lines of code and to accommodate it elsewhere, we keep using mergesort in `names` and manually initialize the rng state in one contrib/ script.

With quadratic case detection, the algorithm becomes a bit more complicated. The existing median-of-three pivot selection system swaps some elements as it selects them which breaks stability. Without the swapping, though, inceraing then decreasing is quadratic, and even with quadratic detection, I expect a detected quadratic case will be much slower than randomized. While quadratic case detection is able to eliminate quadratic runtime in all cases, it does not eliminate pretty bad runtime in more common problematic cases.

Quadratic case detection | Randomized pivot selection
--|--
Impossible to have quadratic runtime | Vanishingly unlikely to have quadratic runtime or even substantially worse than average runtime so long as global rng is not compromised
Does not mutate global random state | Simpler algorithm (fewer parameters to tune, fewer branches to test, fewer lines of code)
Consistent runtime for a given input (helps with benchmarking) | There is no input order with a substantially above average expected runtime
| | Already implemented
| | Supports the possibility of eventually (2.0?) moving MergeSort to SortingAlgorithms.jl

oscardssmith replied:
imo, randomized is much better.

LilithHafner replied:
imo, randomized is slightly better because simplicity is important to me.
Any factors you find especially important in judging it much better?

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

LilithHafner commented:
This depends on workspaces which depend on OffsetArrays. Hold pending availability of OffsetArrays in Sort.jl.

Commit:
Randomize pivot selection

LilithHafner commented:
We're proceeding with workspaces/buffers for sorting without OffsetArrays. We will use a similar approach here as in [radix_sort](45570). It will work but be a bit ugly and have a bit of runtime overhead for certain inputs with offset indexing.

Commit:
fix whitespace

mikmoore commented on stdlib/Random/src/Random.jl:
```diff
@@ -434,4 +434,9 @@ true
+# Randomize quicksort pivot selection. This code is here because of bootstrapping:
+# we need to sort things before we load this standard library.
+# TODO move this into Sort.jl
+Base.Sort.select_pivot(lo::Integer, hi::Integer) = rand(lo:hi)
```
Would it be better to use something other than the global RNG? Perhaps introduce a dedicated stream for sorting operations? I certainly wouldn't anticipate that `sort` would affect the global RNG and it would take me a while to trace the source of the consumption to here. Technically, a user shouldn't care... but that isn't to say they won't.

oscardssmith replied:
I don't think we should try to make the promise that algorithms in `Base`/`Stdlib` won't generate random numbers using the global rng. A ton of stuff needs random numbers (eg `median`, some matrix factorization algorithms (for pivoting), possibly thread scheduling). I think we shouldn't set the expectation that users should care about the global RNG state.

LilithHafner replied:
While it would be nice to not alter the global RNG seed, to maintain our "everything is task scheduling independent, even random numbers" guarantee, which, I believe, is much more important than "sorting doesn't change global RNG state", I think we would need to have a different dedicated stream for each task, which is far too expensive; or we would need to branch a new stream from the global RNG without altering the global RNG's state, which could introduce very hard to anticipate correlations between the randomness of sorting and of other generations akin to but much subtler (and less likely to come up) than https://github.com/JuliaLang/julia/issues/6573

In general, I think everything should use the global RNG, and if someone wants a stream nothing else touches, then they are responsible for creating their own dedicated stream.

There is a warning at the top of https://docs.julialang.org/en/v1/stdlib/Random/ that is somewhat relevant, though it doesn't directly apply to this.

LilithHafner replied:
Beat me to it :) I totally agree.

gaurav-arya replied:
This seems to have caused a bug in Zygote: https://github.com/FluxML/Zygote.jl/issues/1351. Possibly this should just be a pure Zygote fix, but I thought it would be good to be aware that this has the potential to cause some bugs in cases where this function was assumed not to affect the RNG state.

gaurav-arya replied:
A hidden RNG doesn't seem ideal to me because it feels like a good invariant is that a) the global RNG's initial state, and b) any user-provided RNGs make a program'd output deterministic. Hidden RNGs break that; now the user needs to know about them too.

Of course, the output of `sort!` is ultimately deterministic, which helps. But the computation path still isn't: stick an accumulator into the `by` function and now there's observable behavior dependent on the hidden RNG.

That being said, I really have no idea what the best solution to #48230 is, assuming people agree it's a problem. Perhaps all the call points within the internals (REPL etc.) as well as any "ex-internals" (Zygote etc.) would all need to be fixed to use a sorting algorithm that guarantees no random calls, but that's a sizable effort. 

I'll leave it up to people who really know what they're talking about to decide, but just trying to point out the issues as I see it here:)

StefanKarpinski replied:
Forking a new RNG off of the default RNG would be a possibility and just feeding it through all the internal `sort!` calls. We don't really need good randomness for sorting, just something that won't accidentally do something with a pattern that matches the data.

gaurav-arya replied:
Modifying the internal calls does seem reasonable. And some effort would then need to made to find cases like the Zygote one in the ecosystem.

As for how to modify the calls, forking the RNG and using that as you suggest seems reasonable (I don't believe the `sort!` currently accepts user provided RNGs, so that would also need to be fixed, which should probably happen anyway), although just using e.g. `MergeSort` and adding a documented guarantee that the algorithm is purely deterministic seems simpler to me? That is, unless the speed of `QuickerSort` is needed for the internal calls.

LilithHafner replied:
Using `hash(hi) % (hi-lo+1) + lo` instead of `rand(lo:hi)` would be the easiest solution and should be robust against pretty much all non-malicious input. In 1.8 we also had pathological (and more easily constructed) quadratic cases so this would not be a regression.

This would also remove sorting's interaction with https://github.com/JuliaLang/julia/pull/48069 entirely.

oscardssmith replied:
that would be reasonable.

gaurav-arya replied:
That seems reasonable to me? I still think something should be done about the "does sort! semantically guarantee determinism" question because if the answer is no then it's internal usage in various places is incorrect.

But perhaps it would be good in any case to have a version cycle to mull over that? And do the quick fix you suggest for now, since it's not a regression as you say.

oscardssmith replied:
it depends what you mean by determinism.

gaurav-arya replied:
What I mean in my usage is: can `sort!` affect the global RNG state? (As it should, if it makes calls to rand and does not use a hidden RNG which is bad for other reasons.)

Likely, the answer should be "yes, it can", even if it doesn't in 1.8 and now won't in 1.9 with the reversion to hash. In which case, the semantics of the repl+zygote are messed up. So their calls to `sort!` need the answer to the above question to be a guaranteed no (either by using a "deterministic" alg, or by using a special user-provided RNG to a "random" alg).

LilithHafner replied:
> it's internal usage in various places is incorrect.

Do you have an example?

> In 1.8 we also had pathological (and more easily constructed) quadratic cases

For example, the "sawtooth" ordering:

```julia
julia> n = 10^6;

julia> x = vcat(3n:4n,2n:3n,n:2n,0:n);

julia> @time sort!(shuffle!(x));
  0.477350 seconds

julia> x = vcat(3n:4n,2n:3n,n:2n,0:n);

julia> @time sort!(x);
134.471621 seconds

julia> VERSION
v"1.8.5"
```

gaurav-arya replied:
I don't have a clean MWE, but there are two clunkier examples in https://github.com/JuliaLang/julia/issues/48230 (Zygote is one breakage, and then the repl which is a theoretical one since we probably get a dispatch to insertionsort in practice)

gaurav-arya replied:
But you could imagine it could have a lot of consequences in various non-barebones environments where `sort!` is used incautiously, e.g. maybe in Pluto notebook, for whatever reason a `sort!` is run when processing a cell output (this is purely hypothetical). Ideally we'd want the same code with the same seed to run the same in a these environments, which would be in danger if `rand` is introduced to `sort!` without carefully examining such cases

LilithHafner replied:
The default rng should not be used for reproducible random streams.

gaurav-arya replied:
It's very common practice to begin a piece of code with `Random.seed!(...)` to make it reproducible. Faulty a strategy as it may be, it would be confusing if calls to `sort!` made by non user -facing code broke this. That's why I think those calls should eventually be changed to something which can be relied not to affect the global RNG.


In any case, I don't want to debate this point too much here since I'm not pushing for this to affect 1.9.

Commit:
Merge branch 'master' into stabilize-quicksort

Commit:
style

Commit:
Seed rng before generating precompile statements

This fixes a hang where string interpolation in "precompile(Tuple{typeof(show), $IO, $T})\n"
uses sorting which uses rand(lo:hi) which uses an itterated approach which runs forever because
every random UInt64 is 0.

TODO find a better place for this initialization to prevent this bug from cropping up again
perhaps initialize the moment rand is defined (that may be hard).

LilithHafner committed:
fix doctests (1/2)

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

LilithHafner commented on contrib/generate_precompile.jl:
```diff
@@ -19,6 +19,12 @@ CTRL_C = '\x03'
 UP_ARROW = "\e[A"
 DOWN_ARROW = "\e[B"
 
+task = current_task()
+task.rngState0 = 0xb97cb0d700798446
+task.rngState1 = 0xfee2eb4d95cae2c1
+task.rngState2 = 0xa19fbca182930be0
+task.rngState3 = 0xc48c471658f3422e
+
```
Could this go somewhere else?

LilithHafner commented on stdlib/Random/src/Random.jl:
```diff
@@ -434,4 +434,9 @@ true
+# Randomize quicksort pivot selection. This code is here because of bootstrapping:
+# we need to sort things before we load this standard library.
+# TODO move this into Sort.jl
+Base.Sort.select_pivot(lo::Integer, hi::Integer) = rand(lo:hi)
```
This introduces a method overwrite warning during the build

LilithHafner commented on stdlib/Random/docs/src/index.md:
```diff
@@ -168,9 +168,9 @@ julia> rand(Die, 3)
 
 julia> a = Vector{Die}(undef, 3); rand!(a)
 3-element Vector{Die}:
+ Die(9)
```
This change is due to the unfortunate (but not buggy) behavior @mikmoore pointed out, in conjunction with these doctests not heeding the [warning](https://docs.julialang.org/en/v1/stdlib/Random/#Random-Numbers) not to depend on specific random sequences—a warning that is hard to follow in this case.

This PR also breaks doctests in SparseArrays for the same reason. I'm not sure how to edit SparseArrays in a PR to JuliaLang/julia

N5N3 replied:
We can open a PR on `SparseArrays.jl` side to make its doctest independent from the global RNG state by fixing seed manually.
Then it would pass with/without changes here.

LilithHafner replied:
My suspicion was right. Documenter.jl mutates the random state on this PR by converting REPL output to a string. The results are inconsistent between evaluation in Main and running doctests because the rng mutations of `show` on this PR depend on the number of exported names of parent modules.

Cross-ref some other show-related doctest issues https://github.com/JuliaDocs/Documenter.jl/issues/1631 & slightly less relevant https://github.com/JuliaDocs/Documenter.jl/issues/1580.

While it is technically fair game to mutate random state in `show`, it feels like a lot to ask the folks at Documenter to deal with just so that we can randomize pivot selection.

@mortenpi, do you think it is reasonable for Documenter's doctests to accommodate `show` mutating default rng state? Specifically, showing a type (or anything printed with a type) that is a part of a module that has a lot of exported names will mutate the global random state based on those list(s) of exported names.

In the interim, I have introduced a hack to work around this issue by suppressing rng state mutation in `names`.

<!--
Replacing
```
value === nothing || Base.invokelatest(show, IOContext(buf, :limit => true), MIME"text/plain"(), value)
```
with
```
value === nothing || if VERSION >= v"1.9.0-DEV"
    t = current_task()
    x = t.rngState0, t.rngState1, t.rngState2, t.rngState3
    Base.invokelatest(show, IOContext(buf, :limit => true), MIME"text/plain"(), value)
    t.rngState0, t.rngState1, t.rngState2, t.rngState3 = x
else
    Base.invokelatest(show, IOContext(buf, :limit => true), MIME"text/plain"(), value)
end
```
[here](https://github.com/JuliaDocs/Documenter.jl/blob/6c8679bac445ab1c656744075b13a57835081cc3/src/DocTests.jl#L339) should fix this issue, but there ought to be a better way.
-->

<details>

<summary> Here is the stacktrace of where select_pivot gets called by Documenter when running doctests: </summary>

```
  [2] select_pivot(lo::Int64, hi::Int64)
    @ Main.__doctest__named__Die ./none:3
  [3] partition!(t::Vector{Symbol}, lo::Int64, hi::Int64, o::Base.Order.ForwardOrdering, v::Vector{Symbol}, rev::Bool)
    @ Base.Sort ./sort.jl:544
  [4] sort!(v::Vector{Symbol}, lo::Int64, hi::Int64, a::PartialQuickSort{Missing, Missing}, o::Base.Order.ForwardOrdering, t::Vector{Symbol}, swap::Bool, rev::Bool)
    @ Base.Sort ./sort.jl:575
  [5] sort!
    @ ./sort.jl:574 [inlined]
  [6] sort!
    @ ./sort.jl:747 [inlined]
  [7] sort!
    @ ./sort.jl:859 [inlined]
  [8] #sort!#8
    @ ./sort.jl:914 [inlined]
  [9] sort!
    @ ./sort.jl:907 [inlined]
 [10] #names#8
    @ ./reflection.jl:102 [inlined]
 [11] names
    @ ./reflection.jl:102 [inlined]
 [12] make_typealias(x::Type)
    @ Base ./show.jl:600
 [13] show_typealias(io::IOContext{IOBuffer}, x::Type)
    @ Base ./show.jl:781
 [14] _show_type(io::IOContext{IOBuffer}, x::Type)
    @ Base ./show.jl:946
 [15] show(io::IOContext{IOBuffer}, x::Type)
    @ Base ./show.jl:941
 [16] print(io::IOContext{IOBuffer}, x::Type)
    @ Base ./strings/io.jl:35
 [17] showarg
    @ ./show.jl:2953 [inlined]
 [18] array_summary
    @ ./show.jl:2896 [inlined]
 [19] summary
    @ ./show.jl:2893 [inlined]
 [20] show(io::IOContext{IOBuffer}, #unused#::MIME{Symbol("text/plain")}, X::Vector{Main.__doctest__named__Die.Die})
    @ Base ./arrayshow.jl:368
 [21] #invokelatest#2
    @ ./essentials.jl:801 [inlined]
 [22] invokelatest
    @ ./essentials.jl:798 [inlined]
 [23] result_to_string
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:338 [inlined]
 [24] checkresult(sandbox::Module, result::Documenter.DocTests.Result, meta::Dict{Symbol, Any}, doc::Documenter.Documents.Document)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:318
 [25] eval_repl(block::Documenter.DocTests.MutableMD2CodeBlock, sandbox::Module, meta::Dict{Symbol, Any}, doc::Documenter.Documents.Document, page::String)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:247
 [26] doctest(ctx::Documenter.DocTests.DocTestContext, block_immutable::Documenter.Utilities.Markdown2.CodeBlock)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:185
 [27] (::Documenter.DocTests.var"#1#2"{Documenter.DocTests.DocTestContext})(node::Documenter.Utilities.Markdown2.CodeBlock)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:111
 [28] walk(f::Documenter.DocTests.var"#1#2"{Documenter.DocTests.DocTestContext}, node::Documenter.Utilities.Markdown2.CodeBlock)
    @ Documenter.Utilities.Markdown2 ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Markdown2.jl:297
 [29] walk(f::Function, nodes::Vector{Documenter.Utilities.Markdown2.MarkdownBlockNode})
    @ Documenter.Utilities.Markdown2 ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Markdown2.jl:306
 [30] walk
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Markdown2.jl:299 [inlined]
 [31] doctest
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:108 [inlined]
 [32] doctest
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:66 [inlined]
 [33] doctest(blueprint::Documenter.Documents.DocumentBlueprint, doc::Documenter.Documents.Document)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:50
 [34] runner(#unused#::Type{Documenter.Builder.Doctest}, doc::Documenter.Documents.Document)
    @ Documenter.Builder ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Builder.jl:213
 [35] dispatch(#unused#::Type{Documenter.Builder.DocumentPipeline}, x::Documenter.Documents.Document)
    @ Documenter.Utilities.Selectors ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Selectors.jl:170
 [36] #2
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Documenter.jl:267 [inlined]
 [37] cd(f::Documenter.var"#2#3"{Documenter.Documents.Document}, dir::String)
    @ Base.Filesystem ./file.jl:112
 [38] makedocs(; debug::Bool, format::Documenter.Writers.HTMLWriter.HTML, kwargs::Base.Pairs{Symbol, Any, NTuple{11, Symbol}, NamedTuple{(:build, :modules, :clean, :doctest, :linkcheck, :linkcheck_ignore, :strict, :checkdocs, :sitename, :authors, :pages), Tuple{String, Vector{Module}, Bool, Bool, Bool, Vector{String}, Bool, Symbol, String, String, Vector{Any}}}})
    @ Documenter ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Documenter.jl:266
 [39] top-level scope
    @ ~/.julia/dev/julia/doc/make.jl:296
in expression starting at /Users/x/.julia/dev/julia/doc/make.jl:296
```

</details>

mortenpi replied:
How would you accommodate this in Documenter? Do you mean something like storing the RNG state, running `show`, and then resetting the state, as if `show` had not been called?

As an alternative, would using [doctest regex filters](https://juliadocs.github.io/Documenter.jl/dev/man/doctests/#Filtering-Doctests) be an option to work around this issue?

LilithHafner replied:
> storing the RNG state, running `show`, and then resetting the state

That's the only way I can think of to do it. It's not perfect because if someone copied a show + RNG example from a REPL to a doctest it might behave differently, albeit consistently differently, as a doctest. Empirically, that edge case only comes up twice in all of JuliaLang/julia's CI. In those uncommon cases, regex filters or `doctest(fix=true)` should work as workarounds.

LilithHafner replied:
x-ref https://github.com/JuliaDocs/Documenter.jl/pull/1869

LilithHafner replied:
I like to think of sorting as pretty cheap, but yes, it can't hurt!

LilithHafner reviewed (commented):
Most CI tests pass, I expect performance is still good (but have not verified), and all the added perks—stability, accommodating broken user provided `lt`, and making pathological inputs much harder to craft—are present. All this in fewer lines of source code and more lines of comments and test code.

While this isn't quite ready to merge, I think it is once again ready for review. I'd especially appreciate recommendations about how to deal with the consequences of pivot randomization.

Commit:
revert e70ae49ac fix doctests (1/2)

Commit:
fix doctests (2/2); hack names to not mutate rng state

Commit:
Use a better hack to fix doctests

Rather than saving and restoring rng state, we simply use MergeSort.
This is a bit unfortunate as it is the _only_ use of MergeSort in Base or stdlibs,
but it is not a regression because we previously sorted symbols with
MergeSort by default.

LilithHafner committed:
Avoid method overwrite warning during build

Commit:
introdce an easter egg

Commit:
Use an even better hack to avoid breaking docstrings

It's so good it's not even a hack. The problems come from an unfortunately placed
call to sort, but we don't actually need to sort in those cases.

LilithHafner committed:
Merge branch 'JuliaLang:master' into stabilize-quicksort

LilithHafner committed:
Merge branch 'JuliaLang:master' into stabilize-quicksort

oscardssmith commented:
This mostly looks good, but there are fairly substantial regressions for inputs that are either fully forward or reverse sorted that might be good to fix with an issorted check.

LilithHafner commented:
I suppose that wouldn't be too hard to add. I'd also add an undocumented keyword argument to suppress that check to avoid doing it twice in `AdaptiveSort`.

I just noticed that the regression for long vectors of vectors also applies to long vectors of strings. There's also a 22% regression here:
```julia
v = [randstring(4) for _ in 1:1_000_000]; @elapsed sort!(v)
```

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

Commit:
Add presorted check

Commit:
tweak threshold because quicksort is a bit better now

Commit:
fixup

LilithHafner commented:
I've added presorted and reverse-sorted checks and updated [the benchmarks](https://github.com/JuliaLang/julia/pull/45222#issuecomment-1188275456). This makes the graphs look real pretty, but doesn't help with almost presorted or almost reverse-presorted inputs.

I want to be clear that this algorithm change is not a strict improvement. It is and almost certainly will remain a performance regression in multiple meaningful cases. We will need to make a value judgment about whether the improvements (performance and otherwise) outweigh the regressions.

Commit:
test #32675

Commit:
tighten presorted check on partial sort

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

Commit:
tighten reverse-sorted check further to ensure stability

Commit:
test stable reverse short circuit

Commit:
organize sorting tests sligthly & test that PartialQuickSort is stable

LilithHafner committed:
Clarify comment

LilithHafner commented:
Bump. If possible, I'd like to get this into 1.9 because it redefines the AdaptiveSort type which was introduced in 1.9.

oscardssmith commented on base/sort.jl:
```diff
@@ -473,29 +459,34 @@ Characteristics:
+  * *not in-place* in memory.
   * *divide-and-conquer*: sort strategy similar to [`MergeSort`](@ref).
-  * *good performance* for large collections.
+  * *good performance* for most large collections.
```
almost all

LilithHafner replied:
Agreed

oscardssmith commented on base/sort.jl:
```diff
@@ -528,75 +532,92 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
+# This method is redefined to rand(lo:hi) in Random.jl
+# We can't use rand here because it is not available in Core.Compiler and
+# because rand is defined in the stdlib Random.jl after sorting is used in Base.
+select_pivot(lo::Integer, hi::Integer) = midpoint(lo, hi)
```
@timholy will this cause an invalidation problem?

oscardssmith replied:
also, does the compiler need a slightly better algorithm (e.g. median of 3)? I'm not sure that the compiler won't hit the edge cases that would turn this quadratic (and we've been spending a lot of effort removing quadratic compiler problems).

LilithHafner replied:
All these shenanigans should be unnecessary after https://github.com/JuliaLang/julia/pull/46679 because we will then be able to simply depend on Random.

LilithHafner replied:
#47066 alone (not needing all of #46679) should make "does the compiler need a slightly better algorithm (e.g. median of 3)?" outdated.

LilithHafner replied:
As for invalidations, I'm no Tim Holy, but I injected the following snippet into Random.jl right before the method redefinition
```julia
let
    function backedges(f::Function)
        s = Set{Base.MethodInstance}()
        for m in methods(f)
            for mi in m.specializations
                if mi !== nothing
                    backedges!(s, mi)
                end
            end
        end
        s
    end

    function backedges!(s::Set{Base.MethodInstance}, m::Base.MethodInstance)
        if isdefined(m, :backedges)
            for mi in m.backedges
                if mi ∉ s
                    push!(s, mi)
                    backedges!(s, mi)
                end
            end
        end
        s
    end

    @show length(backedges(Base.Sort.select_pivot))
end
```
and got the result
```
length(backedges(Base.Sort.select_pivot)) = 213
```
So there are 213 invalidations at that time. Luckely, Random is loaded before precompilation statements are executed. With and without the redefinition in Random.jl I got about* 3670 direct and indirect backedges from `Base.Sort.select_pivot` which makes me think that the invalidated methodinstances get recompiled later on in the build process so there should be no impact on latency after the build finishes.

*there is a bit of variability because of my manual interaction with the REPL scrollback buffer which causes various methods to be compiled)

I think this result resolves this conversation.

oscardssmith commented on contrib/generate_precompile.jl:
```diff
@@ -19,6 +19,12 @@ CTRL_C = '\x03'
 UP_ARROW = "\e[A"
 DOWN_ARROW = "\e[B"
 
+task = current_task()
```
why? IMO, this needs a comment to explain.

LilithHafner replied:
Because rng state is not initialized at this point and we need it for sorting. I decided to remove what I think is an unnecessary sorting operation later in the file rather than keep the rng initialization.

LilithHafner replied:
#46278 reintroduced another use of sorting and therefore Random during bootstrapping (via `names`). The number of these problems that have come up now (3) makes me want to fix the issue at the source: we fail to initialize RNG state when the Random stdlib is first loaded. I've patched that in 5b8242144522 with a comment that should hopefully clarify the purpose of those lines.

It would be possible to do this in fewer lines by putting a call to `__init__()` in `stdlib/Random/src/RNG.jl`, but as this code is strictly for bootstrapping, I think it is better to keep in `base/sysimg.jl`.

Commit:
remove rng initialization and the unnecessary sort that requires it

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

Commit:
fix merge conflict with 46278

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

Commit:
Update docstring thanks @oscardssmith

LilithHafner commented:
Note: consider https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/19. Not the implementation, but the concept. If the pivot appears repeatedly, utilize that.

Perhaps have a partition end condition based on equality with pivot rather than overlapping indicies.

LilithHafner commented:
Upon further consideration, I can't think of a way that special case optimization is compatible with the partitioning approach of this PR. I think it's best to proceed without that addition and add it later if possible and advantagous.

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

Commit:
better deterministic pivot selction

Commit:
fixup. using Base: hash

LilithHafner committed:
Merge branch 'master' into stabilize-quicksort

LilithHafner commented:
I think that I've addressed all review comments so far. If it's viable, it would be nice to get this into 1.9.

petvana commented on base/sort.jl:
```diff
@@ -533,75 +537,92 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
+# This method is redefined to rand(lo:hi) in Random.jl
+# We can't use rand here because it is not available in Core.Compiler and
+# because rand is defined in the stdlib Random.jl after sorting is used in Base.
+select_pivot(lo::Integer, hi::Integer) = typeof(hi-lo)(hash((lo, hi)) % (hi-lo+1)) + lo
```
```suggestion
select_pivot(lo::Integer, hi::Integer) = typeof(hi-lo)(hash(lo) % (hi-lo+1)) + lo
```
It seems unnecessary to hash tuple of both `(lo,hi)` because of the following mod operation (length `(hi-lo+1)` changes over time). Thus, the pivot selection can be sped up, as demonstrated using v1.8.1
``` julia
julia> using BenchmarkTools

julia> select_pivot(lo::Integer, hi::Integer) = typeof(hi-lo)(hash((lo, hi)) % (hi-lo+1)) + lo
select_pivot (generic function with 1 method)

julia> @btime select_pivot(a,a+b) setup=((a,b) = rand(1:10^6, 2));
  7.441 ns (0 allocations: 0 bytes)

julia> select_pivot_fast(lo::Integer, hi::Integer) = typeof(hi-lo)(hash(lo) % (hi-lo+1)) + lo
select_pivot_fast (generic function with 1 method)

julia> @btime select_pivot_fast(a,a+b) setup=((a,b) = rand(1:10^6, 2));
  3.790 ns (0 allocations: 0 bytes)
```

LilithHafner replied:
You have a fast computer! I get 13 and 10 ns respectively. But that is still a clear improvement and your reasoning w.r.t. the mod operation is sound. Thanks!

LilithHafner replied:
That's surprising. I have `CPU: 4 × Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz` MacBook (only two physical cores) which should theoretically be 1.6x faster, not 1.7-2.6x slower. I've shared my system benchmarks [here](https://github.com/IanButterworth/SystemBenchmark.jl/issues/8#issuecomment-1197475145).

LilithHafner committed:
Optimize deterministic select_pivot

Co-authored-by: Petr Vana <petvana@centrum.cz>

Commit:
update radix sort heuristic because quicksort is faster and the primary competition

LilithHafner commented:
Bump <3

oscardssmith commented on stdlib/Random/src/Random.jl:
```diff
@@ -434,4 +434,10 @@ true
 
+# Randomize quicksort pivot selection. This code is here because of bootstrapping:
+# we need to sort things before we load this standard library.
+# TODO move this into Sort.jl
```
now that the compiler's sort is separated, is this still necessary?

LilithHafner replied:
Yes. If someone filters out the Random stdlib, I'd like to use hash-based pivot selection rather than error trying to use `rand`.

LilithHafner replied:
If sorting were split into a small portion in base that works alone and a larger stdlib that pirates base's sorting to make it run a bit faster and depends on Random, then this code would be able to move to the sorting stdlib, so the TODO is still appropriate.

LilithHafner replied:
And if you mean can we remove that redefinition entirely, yes we can. But it would come at a performance penalty (likely negligible) and at the cost of allowing pathological inputs to be intentionally created.

oscardssmith reviewed (approved):
Looks good!

StefanKarpinski commented:
> Now that every sorting algorithm in Base's sorting policy except InsertionSort uses scratch space, it becomes even more valuable to effectively pass scratch-space vectors around. For example, in this PR, `fpsort!` results in redundant scratch space allocations.

The way you explain this suggests to me that a better keyword for that PR would have been something like `scratch` since scratch space is the term you use to talk about it 😄

LilithHafner commented:
c.f. https://github.com/JuliaLang/julia/pull/45222#issuecomment-1189352106

> I want to be clear that this algorithm change is not a strict improvement. It is and almost certainly will remain a performance regression in multiple meaningful cases. We will need to make a value judgment about whether the improvements (performance and otherwise) outweigh the regressions.
